# Finding Best Algorithm For Our Dataset

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#--------------------------------------------------------

from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.svm import SVC

#--------------------------------------------------------

from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score


I am using 2 different Datasets.

Scaled Data And Non-Scaled Data.

Reason is Few Algorithms works on Decision Trees and Few Works on Distances like Eucladian Distance.

Decision Tree based Algorithms Does not require scaled data, hence we will use Non scaled data from RandomForest and Xgboost.

And for all other algorithms we will use standardized data.

In [16]:
df_raw = pd.read_csv("C://Users//admin//Desktop//Employee_Promotion//Data_Preprocessing//raw_data.csv")
df_scaled = pd.read_csv("C://Users//admin//Desktop//Employee_Promotion//Data_Preprocessing//scaled_data.csv")

df_test_raw = pd.read_csv("C://Users//admin//Desktop//Employee_Promotion//Data_Preprocessing//raw_test_data.csv")
df_test_scaled = pd.read_csv("C://Users//admin//Desktop//Employee_Promotion//Data_Preprocessing//scaled_test_data.csv")

In [13]:
df_raw.head()

avg_training_score  awards_won?  KPIs_met >80%  previous_year_rating  \
0                  49            0              1                     5   
1                  60            0              0                     5   
2                  50            0              0                     3   
3                  50            0              0                     1   
4                  73            0              0                     3   

   department_Operations  department_Procurement  \
0                      0                       0   
1                      1                       0   
2                      0                       0   
3                      0                       0   
4                      0                       0   

   department_Sales & Marketing  department_Technology  department_other  \
0                             1                      0                 0   
1                             0                      0                 0   
2                             1                      0                 0   
3                             1                      0                 0   
4                             0                      1                 0   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [17]:
df_scaled.head()

avg_training_score  awards_won?  KPIs_met >80%  previous_year_rating  \
0           -1.075931    -0.154018       1.356878              1.395766   
1           -0.253282    -0.154018      -0.736986              1.395766   
2           -1.001145    -0.154018      -0.736986             -0.250651   
3           -1.001145    -0.154018      -0.736986             -1.897069   
4            0.718939    -0.154018      -0.736986             -0.250651   

   department_Operations  department_Procurement  \
0              -0.510993                -0.38696   
1               1.956975                -0.38696   
2              -0.510993                -0.38696   
3              -0.510993                -0.38696   
4              -0.510993                -0.38696   

   department_Sales & Marketing  department_Technology  department_other  \
0                      1.501543              -0.386960         -0.382397   
1                     -0.665982              -0.386960         -0.382397   
2                      1.501543              -0.386960         -0.382397   
3                      1.501543              -0.386960         -0.382397   
4                     -0.665982               2.584249         -0.382397   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [19]:
x_raw = df_raw.drop(['is_promoted'] , axis = 1)
y_raw = df_raw.is_promoted

In [20]:
x_scaled = df_scaled.drop(['is_promoted'] , axis = 1)
y_scaled = df_scaled.is_promoted

In [23]:
from sklearn.model_selection import train_test_split

x_train_raw , x_test_raw , y_train_raw , y_test_raw = train_test_split(x_raw , y_raw , test_size = 0.30 , random_state = 2)

x_train_scaled , x_test_scaled , y_train_scaled , y_test_scaled = train_test_split(x_scaled , y_scaled , test_size = 0.30 , random_state = 2)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### RandomForestClassifier

In [24]:
classifier = RandomForestClassifier()
classifier.fit(x_train_raw , y_train_raw)
print(classifier.score(x_test_raw , y_test_raw))

scores = cross_val_score(classifier , x_raw , y_raw , cv = 5)
print(scores.mean())

0.9395487441464453
0.9399540351728202


#### Confusion Matrix

In [25]:
y_pred = classifier.predict(x_test_raw)
cm = confusion_matrix(y_test_raw , y_pred)
cm

array([[14970,    65],
       [  929,   479]], dtype=int64)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### XGBoost

In [26]:
xgb = XGBClassifier()
xgb.fit(x_train_raw , y_train_raw)
print(xgb.score(x_test_raw , y_test_raw))
scores = cross_val_score(xgb , x_raw , y_raw , cv = 5)
print(scores.mean())

0.940825883354619
0.9413042086936049


#### Confusion Matrix

In [27]:
y_pred = xgb.predict(x_test_raw)
cm = confusion_matrix(y_test_raw , y_pred)
cm

array([[14999,    36],
       [  937,   471]], dtype=int64)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### SVM

In [28]:
svm = SVC(kernel = 'linear')
svm.fit(x_train_scaled , y_train_scaled)
print(svm.score(x_test_scaled , y_test_scaled))
scores = cross_val_score(svm , x_scaled , y_scaled , cv = 5)
print(scores.mean())

0.9211822660098522
0.9217450165529406


#### Confusion Matrix

In [29]:
y_pred = svm.predict(x_test_scaled)
cm = confusion_matrix(y_test_scaled , y_pred)
cm

array([[15035,     0],
       [ 1296,   112]], dtype=int64)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

# KNN

In [31]:
knn = KNeighborsClassifier()
knn.fit(x_train_scaled , y_train_scaled)
print(knn.score(x_test_scaled , y_test_scaled))
scores = cross_val_score(knn , x_scaled , y_scaled , cv = 5)
print(scores.mean())

0.9348659003831418
0.9361042561857993


#### Confusion Matrix

In [32]:
y_pred = knn.predict(x_test_scaled)
cm = confusion_matrix(y_test_scaled , y_pred)
cm

array([[14886,   149],
       [  922,   486]], dtype=int64)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

# LogisticRegression

In [33]:
log = LogisticRegression()
log.fit(x_train_scaled , y_train_scaled)
print(log.score(x_test_scaled , y_test_scaled))
scores = cross_val_score(knn , x_scaled , y_scaled , cv = 5)
print(scores.mean())

0.9222769567597153
0.9361042561857993


#### Confusion Matrix

In [34]:
y_pred = log.predict(x_test_scaled)
cm = confusion_matrix(y_test_scaled , y_pred)
cm

array([[14934,   101],
       [ 1177,   231]], dtype=int64)

#### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### Overall Xgboost gives best accuracy with 94% accruracy also it is generalized model

### test score :- 94.08

### cross validarion score :- 94.13

### Model does not need tuning because it is giving good accuracy

### with minimum type1 error and 1 type2 error.

### Hence Xgboost will be selected for prediction.